Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -184,25 +184,62 @@ namespace
         }
     }
 
+    void SetXRGamepadObjectData(Napi::Object& jsInputSource, Napi::Object& jsGamepadObject, xr::System::Session::Frame::InputSource& inputSource)    
+    {
+        auto env = jsInputSource.Env();
+        //Set Gamepad Object
+
+        auto gamepadButtons = Napi::Array::New(env, inputSource.GamepadObject.Buttons.size());
```

Comment:
nit: Just as a general rule, if this is being done per-frame, it might be better to create a "scratch" object that is used repeatedly because JavaScript allocations tend to be _much_ more computationally expensive than seems entirely reasonable. @bghgary can probably comment more informedly on this, but I believe this holds true when allocating objects through N-API. I know that NativeXr.cpp currently does a lot of per-frame JS allocation, so it might be preferable to just deal with all of those at once in a perf pass rather than trying to keep up with them here, so I don't think this needs to be a significant concern right now. It's just something to be aware of.

Reply:
+1 with Justin's suggestion here. It seems like we grab a reference to the javascript gamepad that we then use throughout the Babylon.js codebase. It seems like theres not an expectation for gamepad buttons/state to be immutable.

That going said, if WebXR declares any of these items as immutable (like immutable sets with plane detection), we may be forced to conform to recreating javascript objects every frame. I just wanted to call out that we may only be able to introduce these improvements case by case.

---
In reply to: [523280642](https://github.com/BabylonJS/BabylonNative/pull/497#discussion_r523280642) [](ancestors = 523280642)